src/proxy/h2_client.rs

```diff
@@ -0,0 +1,344 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This code draws major inspiration from 2 Apache-2.0 licensed codebases:
+// * https://github.com/cloudflare/pingora/blob/main/pingora-core/src/protocols/http/v2/client.rs
+// * https://github.com/hyperium/hyper/blob/master/src/proto/h2/client.rs
+
+use crate::config;
+use crate::proxy::Error;
+use bytes::Buf;
+use bytes::Bytes;
+use futures_core::ready;
+use h2::client::{Connection, SendRequest};
+use h2::{Reason, SendStream};
+use http::Request;
+use std::io::Cursor;
+use std::pin::Pin;
+use std::sync::atomic::{AtomicBool, AtomicU16, Ordering};
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use std::time::Duration;
+use tokio::io::{AsyncRead, AsyncWrite, ReadBuf};
+use tokio::net::TcpStream;
+use tokio::sync::oneshot;
+use tokio::sync::watch::Receiver;
+use tokio_rustls::client::TlsStream;
+use tracing::{debug, error, trace, warn};
+
+// H2Stream represents an active HTTP2 stream. Consumers can only Read/Write
+pub struct H2Stream {
+    send_stream: h2::SendStream<SendBuf>,
+    recv_stream: h2::RecvStream,
+    buf: Bytes,
+    active_count: Arc<AtomicU16>,
+}
+
+impl H2Stream {
+    fn write_slice(&mut self, buf: &[u8], end_of_stream: bool) -> Result<(), std::io::Error> {
+        let send_buf: SendBuf = Cursor::new(buf.into());
+        self.send_stream
+            .send_data(send_buf, end_of_stream)
+            .map_err(h2_to_io_error)
+    }
+}
+
+impl Drop for H2Stream {
+    fn drop(&mut self) {
+        let left = self.active_count.fetch_sub(1, Ordering::SeqCst);
+        trace!("dropping h2stream, has {} active streams left", left - 1);
+    }
+}
+
+impl AsyncRead for H2Stream {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        read_buf: &mut ReadBuf<'_>,
+    ) -> Poll<std::io::Result<()>> {
+        if self.buf.is_empty() {
+            self.buf = loop {
+                match ready!(self.recv_stream.poll_data(cx)) {
+                    None => return Poll::Ready(Ok(())),
+                    Some(Ok(buf)) if buf.is_empty() && !self.recv_stream.is_end_stream() => {
+                        continue
+                    }
+                    Some(Ok(buf)) => {
+                        // TODO: Hyper and Go make their pinging data aware and don't send pings when data is received
```

Comment:
GRPC uses ping bytes to estimate "bandwidth delay product" which can be useful for LB down the road.